### PR TITLE
[0698] Migrate courses add course form button

### PR DESCRIPTION
**Pull request by defong**
### Context
Migrate courses add course form button

### Changes proposed in this pull request
Clean up a lot of mixed concepts that occurred from https://github.com/DFE-Digital/teacher-training-api/pull/2471
Fixed up course creation for `further education`
Mostly over using `CourseBasicDetailConcern`

### Guidance to review

The WHOLE course creation journey

Use a provider with 
- no locations
- only one locations
- 2 or more locations

create a course
- requires accrediting bodies
- does not requires accrediting bodies
- level
  - primary
  - secondary
    - one subject
    - two subjects
    - modern languages 
  - further education
- custom data
  - age range
  - dates etc

### Checklist

- [x] Make sure all information from the Trello card is in here
- [x] Attach to Trello card
- [x] Rebased master
- [x] Cleaned commit history
- [x] Tested by running locally

### CODE REVIEW

File: spec/services/course_creation_service_spec.rb [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+      expect(subject.subjects.map(&:id)).to eq([primary_subject.id])
+      expect(subject.course_code).to eq(nil)
+    end
+
+    context "next_available_course_code is true" do
+      subject { described_class.call(course_params: course_params(:primary), provider: provider, next_available_course_code: true) }
+
+      it "create the primary course" do
+        primary_course_params.except("is_send", "sites_ids", "subjects_ids", "course_code").each do |key, value|
+          expect(subject.send(key)).to eq(value)
```
Comment (darokel) [inline spec/services/course_creation_service_spec.rb:23]: Could we use `.public_send` here for convention (and anywhere else)

File: app/services/course_creation_service.rb [hunk 2]
```diff
@@ -0,0 +1,96 @@
+class CourseCreationService
```
Comment (darokel) [inline app/services/course_creation_service.rb:1]: Any service prefixed with `Course` should probably utilise the namespace/directory we have here?

File: spec/services/course_creation_service_spec.rb [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+      "is_send" => "1",
+      "level" => "further_education",
+      "qualification" => "pgde",
+      "start_date" => "September #{recruitment_cycle.year}",
+      "study_mode" => "full_time_or_part_time",
+      "sites_ids" => [site.id],
+    }
+  end
+
+  def provider
```
Comment (darokel) [inline spec/services/course_creation_service_spec.rb:149]: I know we decided on avoiding lets for feature specs specifically but still use them for unit tests. Any reason why we're not using them for these fixtures?

**Review by darokel — COMMENTED** [review]
+1 for refactoring some of this logic. Left a couple of minor comments but looks good overall

File: spec/services/course_creation_service_spec.rb (hunk 3 above)
Comment (defong) [reply spec/services/course_creation_service_spec.rb:149]: ¯\_(ツ)_/¯  cognitive overload

File: spec/services/course_creation_service_spec.rb (hunk 1 above)
Comment (defong) [reply spec/services/course_creation_service_spec.rb]: Done

File: app/services/course_creation_service.rb (hunk 2 above)
Comment (defong) [reply app/services/course_creation_service.rb]: Done

File: spec/services/course_creation_service_spec.rb (hunk 3 above)
Comment (defong) [reply spec/services/course_creation_service_spec.rb]: Done

**Review by darokel — APPROVED** [review]

File: spec/features/publish/course_confirmation_spec.rb [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+private
+
+  def given_i_am_authenticated_as_a_provider_user
+    @user = create(:user, providers: [build(:provider, sites: [build(:site)])])
+    @user.providers.first.courses << create(:course, :with_accrediting_provider)
+    given_i_am_authenticated(user: @user)
+  end
+
+  def when_i_visit_the_confirmation_page
+    confirmation_page.load(provider_code: provider.provider_code, recruitment_cycle_year: Settings.current_recruitment_cycle_year, query: confirmation_params(provider))
```
Comment (darokel) [inline spec/features/publish/course_confirmation_spec.rb:28]: could we break this up, bit long

File: spec/features/publish/courses/new_mordern_languages_spec.rb [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    query = new_course_modern_languages_page_with_query(invalid: with_invalid_query)
+    new_modern_languages_page.load(provider_code: provider.provider_code, recruitment_cycle_year: Settings.current_recruitment_cycle_year, query: query)
+  end
+
+  def and_i_click_continue
+    new_modern_languages_page.continue.click
+  end
+
+  def language
+    @language ||= %i[french english_as_a_second_language_or_other_language german italian japanese mandarin russian spanish modern_languages_other].sample
```
Comment (darokel) [inline spec/features/publish/courses/new_mordern_languages_spec.rb:58]: this one too

File: spec/features/publish/new_course_spec.rb [hunk 6]
```diff
@@ -0,0 +1,243 @@
+require "rails_helper"
+
+feature "new course" do
```
Comment (darokel) [inline spec/features/publish/new_course_spec.rb:3]: we could put this in an `e2e` directory under `/features/publish/` similar to how we did it in register

File: spec/support/page_objects/publish/course_confirmation.rb [hunk 7]
```diff
@@ -0,0 +1,41 @@
+# frozen_string_literal: true
+
+require_relative "../sections/errorlink"
+module PageObjects
```
Comment (darokel) [inline spec/support/page_objects/publish/course_confirmation.rb:4]: spacing here please

File: spec/support/page_objects/publish/courses/new_modern_languages_page.rb [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+module PageObjects
+  module Publish
+    module Courses
+      class NewModernLanguages < PageObjects::Base
+        set_url "/publish/organisations/{provider_code}/{recruitment_cycle_year}/courses/modern-languages/new{?query*}"
+
+        element :languages_fields, '[data-qa="course__languages"]'
+        element :title, '[data-qa="page-heading"]'
+
+        def language_checkbox(name)
```
Comment (darokel) [inline spec/support/page_objects/publish/courses/new_modern_languages_page.rb:12]: Could we place the methods last before the DSLs

File: spec/features/publish/new_course_spec.rb [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+      expected_query_params: course_creation_params,
+    )
+
+    course_creation_params
+  end
+
+  def select_age_range(course_creation_params, next_page:)
+    course_creation_params[:age_range_in_years] = "5_to_11"
+
+    choose("course_age_range_in_years_5_to_11")
```
Comment (darokel) [inline spec/features/publish/new_course_spec.rb:131]: Does this not have a page object?

**Review by darokel — COMMENTED** [review]
👏 nice work on the specs, you've saved us a few painful rabbit holes down the road. I've left a few cosmetic suggestions/questions. Let me know your thoughts and then we can get this in

File: spec/features/publish/course_confirmation_spec.rb (hunk 4 above)
Comment (defong) [reply spec/features/publish/course_confirmation_spec.rb]: done in https://github.com/DFE-Digital/teacher-training-api/pull/2525/commits/77827debc4ed2be38058f60183926ff5264ee34a

File: spec/features/publish/courses/new_mordern_languages_spec.rb (hunk 5 above)
Comment (defong) [reply spec/features/publish/courses/new_mordern_languages_spec.rb]: done in https://github.com/DFE-Digital/teacher-training-api/pull/2525/commits/77827debc4ed2be38058f60183926ff5264ee34a

File: spec/features/publish/new_course_spec.rb (hunk 6 above)
Comment (defong) [reply spec/features/publish/new_course_spec.rb]: done in https://github.com/DFE-Digital/teacher-training-api/pull/2525/commits/77827debc4ed2be38058f60183926ff5264ee34a

File: spec/support/page_objects/publish/course_confirmation.rb (hunk 7 above)
Comment (defong) [reply spec/support/page_objects/publish/course_confirmation.rb:5]: done in https://github.com/DFE-Digital/teacher-training-api/pull/2525/commits/77827debc4ed2be38058f60183926ff5264ee34a

File: spec/support/page_objects/publish/courses/new_modern_languages_page.rb (hunk 8 above)
Comment (defong) [reply spec/support/page_objects/publish/courses/new_modern_languages_page.rb:14]: done in https://github.com/DFE-Digital/teacher-training-api/pull/2525/commits/77827debc4ed2be38058f60183926ff5264ee34a

File: spec/features/publish/new_course_spec.rb (hunk 9 above)
Comment (defong) [reply spec/features/publish/new_course_spec.rb]: done in https://github.com/DFE-Digital/teacher-training-api/pull/2525/commits/77827debc4ed2be38058f60183926ff5264ee34a

**Review by darokel — APPROVED** [review]

**Pull request merged**